common/src/collision.rs

```diff
@@ -0,0 +1,588 @@
+                let node = graph.neighbor(opposite_node, *side);
+                let translated_position = side.reflection() * opposite_position;
+                // the chunk next to the origin chunk
+                Self::add_sub_bb(
+                    &mut bounding_boxes,
+                    ChunkBoundingBox::get_chunk_bounding_box(
+                        node.unwrap(),
+                        start_chunk,
+                        translated_position,
+                        radius,
+                        dimension,
+                    ),
+                );
+                // the chunk on the opposite side of the edge
+                Self::add_sub_bb(
+                    &mut bounding_boxes,
+                    ChunkBoundingBox::get_chunk_bounding_box(
+                        node.unwrap(),
+                        PERPENDICULAR_VERTEX[*side as usize][start_chunk as usize].unwrap(),
+                        translated_position,
+                        radius,
+                        dimension,
+                    ),
+                );
+            }
+
+            // get BB for chunk sharing only a vertex.
+            Self::add_sub_bb(
+                &mut bounding_boxes,
+                ChunkBoundingBox::get_chunk_bounding_box(
+                    opposite_node,
+                    Vertex::from_sides(sides[0], sides[1], sides[2]).unwrap(),
+                    opposite_position,
+                    radius,
+                    dimension,
+                ),
+            );
+        }
+
+        BoundingBox { bounding_boxes }
+    }
+
+    // adds a sub-bounding box to a list if it exists
+    pub fn add_sub_bb(list: &mut Vec<ChunkBoundingBox>, result: Option<ChunkBoundingBox>) {
+        if let Some(x) = result {
+            list.push(x);
+        }
+    }
+
+    pub fn every_voxel_address(&self) -> impl Iterator<Item = VoxelAddress> + '_ {
+        self.bounding_boxes.iter().flat_map(|cbb| {
+            cbb.every_voxel().map(move |index| VoxelAddress {
+                node: cbb.node,
+                chunk: cbb.chunk,
+                index,
+            })
+        })
+    }
+}
+
+// translated_position should be the object position in the node coordinates of the chunk.
+// node can be easily factored out if it stops being convienent.
+impl ChunkBoundingBox {
+    pub fn get_chunk_bounding_box(
+        node: NodeId,
+        chunk: Vertex,
+        translated_position: na::Vector4<f64>,
+        radius: f64,
+        dimension: u8,
+    ) -> Option<Self> {
+        let euclidean_position = {
+            let temp = chunk.chunk_to_node().try_inverse().unwrap() * translated_position;
+            temp.xyz() / temp[3]
+        };
+
+        let mut min_xyz = na::Vector3::<u32>::new(0_u32, 0_u32, 0_u32);
+        let mut max_xyz = na::Vector3::<u32>::new(0_u32, 0_u32, 0_u32);
+
+        // It's important to note that euclidean_position is measured as chunk lengths, and radius is measured in absolute units.
```

Comment:
I don't believe that's an issue. At large radi, the bounding box would translate to a roughly spherical polygonal object, not an ideal cube.

Reply:
Are you sure that's the case? My understanding is that the `chunk_to_node` conversion caps the resulting coordinates within a bounded region in Euclidean space (sphere or ellipsoid, not sure which), and if `radius` expands the (0,0,0) to (1,1,1) cube large enough to intersect the boundary of this region, then the bounding box would become infinite in size.

However, this may be a non-issue in the long term, since this AABB check is meant to be a first pass, and you're also restricting which chunks are used for this test to begin with. It could be a point of confusion, though, so elaborating on this in a comment may still be worthwhile. I still believe performing a conversion here would be worthwhile, since having "radius" mean two separate things seems like it would cause bugs.

Reply:
Some discussion was done on the Discord, which I will repeat here for context: (-0.19, -0.19, -0.19) in chunk coordinates does not map to any physical point in node coordinates, which means that if the `radius` is 0.19 or higher in the current implementation, the resulting bounding box is larger than an ideal cube, so collisions can occur from infinitely far away.

One way to bypass the 0.19-cutoff issue and allow arbitrarily large radii to behave sanely with voxel coordinates is to utilize equidistant surfaces instead of planes for determining whether the player is near a particular voxel. For instance, to get the x-coordinate of `min_xyz`, the answer should be the chunk x-coordinate of the plane that the player is `radius` units away from. I've worked out a formula for this, and it's a solution to a quadratic equation, meaning that it only takes one square root operation to find the x-coordinate. In total, this would be 6 square root operations per call to get_chunk_bounding_box, which can be whittled down to 3 by reusing some calculations.

There are four inputs to this process, all expressed in terms of node coordinates:
- `cube_center`: A normalized vector corresponding to the point in the center of the chunk's cube in the cubic tiling. This corresponds to (1, 1, 1) in chunk coordinates.
- `cube_x`: A normalized vector corresponding to the x-direction from the center of the chunk's cube. This vector should lead from (1, 1, 1) to (0, 1, 1) in chunk coordinates.
- `sinh_entity_radius`: A number corresponding to the hyperbolic sine of the radius of the entity colliding with the world, in units of absolute distance
- `entity_position`: A normalized vector corresponding to the current location of the entity.

The output will be `klein_x`. If, in the Klein-Beltrami ball model centered at `cube_center` with radius 1, we define a plane with x-coordinate `klein_x`, the player should be just touching that plane. This is convenient because in such a projection, voxel coordinates map uniformly to a Euclidean cube with equally spaced voxels, so `max_xyz.x` in `get_chunk_bounding_box` is linearly related to `klein_x`.

The first step is to get relevant numbers:
- `mip_cube_center` = `mip(entity_position, cube_center)`
- `mip_cube_x` = `mip(entity_position, cube_x)`

Then, the equation we want to solve is
```
mip_cube_x * cosh(plane_dist) - mip_cube_center * sinh(plane_dist) == sinh_entity_radius
```
where `plane_dist` is the absolute-coordinates distance from the plane corresponding to `klein_x` to the origin. To discover this equation, one can use the sphere analogy. The question is how far to rotate a unit sphere on a particular axis to bring a particular point to a small circle `sin(entity_radius)` units away from the equator.

Note that `klein_x == tanh(plane_dist)`, so with some hyperbolic trig identities, the equation to solve converts to
```
mip_cube_x / sqrt(1 - klein_x^2) - mip_cube_center * klein_x / sqrt(1 - klein_x^2) == sinh_entity_radius
```

Solving this algebraically yields
```
klein_x = (mip_cube_x * mip_cube_center +- sinh_entity_radius * sqrt(mip_cube_center^2 + sinh_entity_radius^2 - mip_cube_x^2)) / (mip_cube_center^2 + sinh_entity_radius^2)
```
As for the choice of whether to use `+` or `-`, I believe it's as simple as `+` giving you `max_xyz.x` and `-` giving you `min_xyz.x`.

Similar math should get you appropriate `y` and `z` coordinates.

Example code:
```rust
// Set the entity position to arbitrary x, y, and z coordinates, then use the appropriate w coordinate.
let mut entity_position: na::Vector4<f64> = na::Vector4::new(5.0, 4.5, 4.0, 0.);
entity_position[3] = (entity_position[0] * entity_position[0]
    + entity_position[1] * entity_position[1]
    + entity_position[2] * entity_position[2]
    + 1.0)
    .sqrt();
let entity_position = entity_position;

// Arbitrary entity radius
let sinh_entity_radius = 0.3f64;

// For the example's simplicity, we assume the cube is centered at the origin and axis aligned
let cube_center: na::Vector4<f64> = math::origin();
let cube_x: na::Vector4<f64> = na::Vector4::x();

// Computation is here
let mip_cube_center = math::mip(&entity_position, &cube_center);
let mip_cube_x = math::mip(&entity_position, &cube_x);

let klein_x = (mip_cube_x * mip_cube_center
    + sinh_entity_radius
        * (mip_cube_center.powi(2) + sinh_entity_radius.powi(2) - mip_cube_x.powi(2))
            .sqrt())
    / (mip_cube_center.powi(2) + sinh_entity_radius.powi(2));

// Print the result and test that behavior is as expected. The x-coordinate of the result to math::translate_along
// should equal sinh_entity_radius.
println!("{}", klein_x);
println!(
    "{}",
    math::translate_along(&na::Vector3::x_axis(), klein_x.atanh()) * entity_position
);
```

There's still some work needed to figure out how to find `cube_center` and `cube_x`, as well as how to map `klein_x` to a voxel coordinate, but I believe that this should be relatively straightforward.

Reply:
Looking at my explanation again, I realize that this is actually very difficult to implement correctly, especially if the intuition in my explanation isn't 100% there. For that reason, to help out, I made my own implementation, which I've included in this comment here. It definitely needs cleanup, but hopefully, it can help clear things up.

<details>
<summary>See example definition of get_chunk_bounding_box</summary>

```rust
pub fn get_chunk_bounding_box(
    node: NodeId,
    chunk: Vertex,
    translated_position: na::Vector4<f64>,
    radius: f64,
    dimension: u8,
) -> Option<Self> {
    // The following can be precalculated, likely in a better way than this hacky code.
    let cube_center: na::Vector4<f64> = { let temp = chunk.chunk_to_node() * na::Vector4::new(1.0, 1.0, 1.0, 1.0); temp / (-math::mip(&temp, &temp)).sqrt() };
    let sides = chunk.canonical_sides();
    let cube_x = { let temp = sides[0].reflection() * math::origin() - math::origin(); temp / math::mip(&temp, &temp).sqrt() };
    let cube_y = { let temp = sides[1].reflection() * math::origin() - math::origin(); temp / math::mip(&temp, &temp).sqrt() };
    let cube_z = { let temp = sides[2].reflection() * math::origin() - math::origin(); temp / math::mip(&temp, &temp).sqrt() };

    let cube_to_klein = ((-math::mip(&(Side::A.reflection() * Side::B.reflection() * Side::C.reflection() * math::origin()), &math::origin())).acosh() / 2.0).tanh() / 3.0f64.sqrt();

    let sinh_radius = radius.sinh();

    // The following 4 lines should be able to be replaced with a single matrix-vector multiplication.
    let mip_cube_center = math::mip(&translated_position, &cube_center);
    let mip_cube_x = math::mip(&translated_position, &cube_x);
    let mip_cube_y = math::mip(&translated_position, &cube_y);
    let mip_cube_z = math::mip(&translated_position, &cube_z);

    // The following 6 lines have needless code repetition and can be cleaned up. Every pair of lines can be changed to reuse the result of the sqrt part of the calculation.
    let klein_x_min = (mip_cube_x * mip_cube_center - sinh_radius * (mip_cube_center.powi(2) + sinh_radius.powi(2) - mip_cube_x.powi(2)).sqrt()) / (mip_cube_center.powi(2) + sinh_radius.powi(2));
    let klein_x_max = (mip_cube_x * mip_cube_center + sinh_radius * (mip_cube_center.powi(2) + sinh_radius.powi(2) - mip_cube_x.powi(2)).sqrt()) / (mip_cube_center.powi(2) + sinh_radius.powi(2));
    let klein_y_min = (mip_cube_y * mip_cube_center - sinh_radius * (mip_cube_center.powi(2) + sinh_radius.powi(2) - mip_cube_y.powi(2)).sqrt()) / (mip_cube_center.powi(2) + sinh_radius.powi(2));
    let klein_y_max = (mip_cube_y * mip_cube_center + sinh_radius * (mip_cube_center.powi(2) + sinh_radius.powi(2) - mip_cube_y.powi(2)).sqrt()) / (mip_cube_center.powi(2) + sinh_radius.powi(2));
    let klein_z_min = (mip_cube_z * mip_cube_center - sinh_radius * (mip_cube_center.powi(2) + sinh_radius.powi(2) - mip_cube_z.powi(2)).sqrt()) / (mip_cube_center.powi(2) + sinh_radius.powi(2));
    let klein_z_max = (mip_cube_z * mip_cube_center + sinh_radius * (mip_cube_center.powi(2) + sinh_radius.powi(2) - mip_cube_z.powi(2)).sqrt()) / (mip_cube_center.powi(2) + sinh_radius.powi(2));

    // Similarly, the code repetition below can be cleaned up.
    let cube_x_min = 1.0 - klein_x_max / cube_to_klein;
    let cube_x_max = 1.0 - klein_x_min / cube_to_klein;
    let cube_y_min = 1.0 - klein_y_max / cube_to_klein;
    let cube_y_max = 1.0 - klein_y_min / cube_to_klein;
    let cube_z_min = 1.0 - klein_z_max / cube_to_klein;
    let cube_z_max = 1.0 - klein_z_min / cube_to_klein;

    if cube_x_min < 1.0 && cube_x_max > 0.0 && cube_y_min < 1.0 && cube_y_max > 0.0 && cube_z_min < 1.0 && cube_z_max > 0.0 {
        Some(ChunkBoundingBox {
            node,
            chunk,
            min_xyz: na::Vector3::<u32>::new(
                (cube_x_min.max(0.0) * dimension as f64).floor() as u32,
                (cube_y_min.max(0.0) * dimension as f64).floor() as u32,
                (cube_z_min.max(0.0) * dimension as f64).floor() as u32,
            ),
            max_xyz: na::Vector3::<u32>::new(
                (cube_x_max.min(1.0) * dimension as f64).ceil() as u32,
                (cube_y_max.min(1.0) * dimension as f64).ceil() as u32,
                (cube_z_max.min(1.0) * dimension as f64).ceil() as u32,
            ),
            dimension,
        })
    } else {
        None
    }
}
```

</details>

Reply:
Based on conversations in Discord, I've created a cleaner version at https://github.com/DimpyRed/hypermine/pull/16. Some of these changes should likely be reviewed separately by @Ralith, so I created https://github.com/Ralith/hypermine/pull/232 for that purpose, which also makes a minor change to chunk coordinates.